### PR TITLE
Windows

### PR DESCRIPTION
**Pull request by chrismatteson**
Added Windows support for the dnsclient module.

### CODE REVIEW

File: Modulefile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 name    'ghoneycutt-dnsclient'
-version '3.1.1'
+version '3.2.0'
```
Comment (ghoneycutt) [inline Modulefile]: Do not change the version when sending PR's as the maintain will do that when they cut a release. For instance, I might merge another PR before this one and then this version would not be correct.

File: .fixtures.yml [hunk 2]
```diff
@@ -1,7 +1,6 @@
 fixtures:
-  repositories:
-    'stdlib':
-      repo: 'git://github.com/puppetlabs/puppetlabs-stdlib.git'
-      ref: '3.2.0'
+  forge_modules:
```
Comment (ghoneycutt) [inline .fixtures.yml]: please revert this file, as pulling repos without versions will lead to all unpredictability in testing. Add a reference to the powershell module in the same manner that stdlib is listed.

Comment (ghoneycutt) [inline .fixtures.yml]: please revert this file, as pulling repos without versions will lead to all unpredictability in testing. Add a reference to the powershell module in the same manner that stdlib is listed.

File: Gemfile [hunk 3]
```diff
@@ -5,3 +5,4 @@ gem 'puppet', puppetversion
 gem 'puppetlabs_spec_helper', '>= 0.1.0'
 gem 'puppet-lint', '>= 0.3.2'
 gem 'facter', '>= 1.7.0', "< 1.8.0"
+gem 'rspec-core', '3.1.7'
```
Comment (ghoneycutt) [inline Gemfile]: Why was this needed?

File: manifests/init.pp [hunk 4]
```diff
@@ -3,7 +3,7 @@
 # This module manages /etc/resolv.conf and is meant to be included in the
 # common class that applies to all systems.
 #
-class dnsclient                 (
+class dnsclient (
```
Comment (ghoneycutt) [inline manifests/init.pp]: Good catch on style

File: manifests/init.pp [hunk 5]
```diff
@@ -16,29 +16,49 @@
   $resolver_config_file_owner  = 'root',
   $resolver_config_file_group  = 'root',
   $resolver_config_file_mode   = '0644',
-) {
```
Comment (ghoneycutt) [inline manifests/init.pp]: This is as it should be.

File: Modulefile (hunk 1 above)
Comment (chrismatteson) [inline Modulefile]: I'm not sure how that change, and the additions of 7 to all the rhel based operating systems happened.  I thought I forked from master then edited from there, but it looks like I grabbed something unintentionally (I'm still a novice at git, so I hope you'll excuse that.)

File: spec/classes/init_spec.rb [hunk 6]
```diff
@@ -1,169 +1,174 @@
 require 'spec_helper'
```
Comment (ghoneycutt) [inline spec/classes/init_spec.rb]: Please unstage this file. I'm going to redo these spec tests to make them easier to understand and contribute to then will add tests for this PR after rebasing.

File: manifests/init.pp [hunk 7]
```diff
@@ -16,29 +16,49 @@
...
+  ){
 
-  # Validates domain
-  if is_domain_name($domain) != true {
-    fail("Domain name, ${domain}, is invalid.")
+  if $::kernel == 'windows' {
+    if is_array($nameservers) {
+      $stringifynameservers = join($nameservers,',')
+      $testserver = $nameservers[0]
```
Comment (ghoneycutt) [inline manifests/init.pp]: This only pulls the first nameserver from the array instead of setting all of the nameservers. Is that intended?

File: manifests/init.pp [hunk 8]
```diff
@@ -16,29 +16,49 @@
...
+    }
+    else {
+      $stringifynameservers = $nameservers
+      $testserver = $nameservers
+    }
+    exec { 'set dns':
+      command  => "\$interface = find-netroute -RemoteIPAddress ${testserver} | select -expand 'InterfaceAlias'; set-dnsclientserveraddress -interfacealias \$interface -serveraddress {${stringifynameservers}}",
```
Comment (ghoneycutt) [inline manifests/init.pp]: You know of a way to validate powershell in Linux? Thinking that these could go in a template, so that the separate files could be validated.

File: checksums.json [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (ghoneycutt) [inline checksums.json]: Please unstage this file from the commit. This is created by `puppet module build` for the Forge and is not meant to be checked into the repo.

**Comment by ghoneycutt** [comment]
Awesome work!

**Comment by ghoneycutt** [comment]
Please modify the README.md to denote support for Windows

**Comment by ghoneycutt** [comment]
You got me looking at this module some more :)
I modified a few things to bring it up to date with my other modules, so you will need to rebase your commit.

File: manifests/init.pp (hunk 7 above)
Comment (chrismatteson) [inline manifests/init.pp]: Because windows lets you specify a DNS server on every interface, for a multihoned system, I needed to figure out a way to add the DNS entries to the correct interface.  Right now I accomplish this by taking the first server and running it through find-netroute to figure out which interface would be used to get to that IP.  I then assume all the DNS servers should go on that IP.  The best scenario would probably be to evaluate each entry for it's interface and then update accordingly.  I might be able to do that in a template erb file as you recommended, or it might make sense to just write a type/provider (once I decide I want to learn that much ruby...)

File: spec/classes/init_spec.rb (hunk 6 above)
Comment (chrismatteson) [inline spec/classes/init_spec.rb]: I unstaged it, although everything falls into my "unsupported os" bucket now, so CI fails.  Regardless, looks like I've cleaned up all the garbage I accidentally pulled across from the forge and rebased to where you are now.

File: manifests/init.pp (hunk 8 above)
Comment (chrismatteson) [inline manifests/init.pp]: I could replace the command and unless with templates of powershell scripts with a for loop to allow me to handle dns entries on multiple interfaces, and also make this much more readable.  I'll take a whack at that tomorrow.

I also figured out how I could find the registry keys using the proper powershell command.  However it doesn't seem like that really provides a significant advantage, even if I moved the query into a type/provider and just put a registry entry here to avoid use of exec entirely.  Furthermore, I imagine microsoft is more likely to maintain the powershell format going forward than the registry keys.

**Comment by ghoneycutt** [comment]
This module is meant to cater the old style unix way of managing a single resolv.conf. Suggest implementing this functionality in its own module. Thank you for your contribution!
